thorlcr/slave/slavmain.cpp

```diff
@@ -1214,21 +1175,60 @@ class CKJService : public CSimpleInterfaceOf<IKJService>, implements IThreaded,
     void addToKeyManagerCache(CKMContainer *kmc)
     {
         CriticalBlock b(kMCrit);
-        if (numCached == maxCachedKJManagers)
+        if (numKMCached == maxCachedKJManagers)
         {
             CKMContainer &oldest = cachedKMsMRU.item(0);
-            verifyex(removeKeyManager(oldest.queryHandle())); // also removes from cachedKMsMRU
+            auto it = cachedKMs.find(oldest.queryCtx().queryKey());
+            assertex(it != cachedKMs.end());
+            CKMKeyEntry *kme = it->second;
+            verifyex(kme->remove(kmc));
+            if (0 == kme->count())
+                cachedKMs.erase(it);
+            verifyex(cachedKMsMRU.zap(oldest));
+            --numKMCached;
         }
-        CKMKeyEntry *kme = cachedKMs.find(kmc->queryCtx().queryKey());
-        if (!kme)
+        const CLookupKey &key = kmc->queryCtx().queryKey();
+        auto it = cachedKMs.find(key);
+        CKMKeyEntry *kme;
+        if (it == cachedKMs.end())
         {
-            kme = new CKMKeyEntry(kmc->queryCtx().queryKey());
-            cachedKMs.replace(*kme);
+            kme = new CKMKeyEntry(key);
+            cachedKMs.insert({key, kme});
         }
-        kme->push(kmc); // JCSMORE cap. to some max #
-        cachedKMsByHandle.replace(*LINK(kmc));
+        else
+            kme = it->second;
+        kme->push(kmc); // takes ownership. JCSMORE cap. to some max #
         cachedKMsMRU.append(*kmc);
-        ++numCached;
+        ++numKMCached;
+    }
+    void addToFetchContextCache(CFetchContext *fc)
+    {
+        CriticalBlock b(lCCrit);
+        if (numFCCached == maxCachedFetchContexts)
+        {
+            CFetchContext &oldest = cachedFCsMRU.item(0);
+            auto it = cachedFetchContexts.find(oldest.queryKey());
+            assertex(it != cachedFetchContexts.end());
+            CFCKeyEntry *fce = it->second;
+            verifyex(fce->remove(&oldest));
+            if (0 == fce->count())
+                cachedFetchContexts.erase(it);
+            verifyex(cachedFCsMRU.zap(oldest));
```

Comment:
.remove(0) would probably be better.  This is also potentially painful - since if there are 1000 items you will be copying 8K of memory on each remove.  Each time one is removed from the middle it will copy half that.

Could do with a LRU class which solved this once.

Reply:
Hopefully the length is generally much shorter than that - what would you expect it to be?

Reply:
I do expect typical cases to be quite small, but I also I don't expect these MRU movement to be frequent.
It is on key part open / fetch part open, they should remain open and out of cache for as long as the subgraph is running, and re-enter the cache when done.

If there were high numbers, and parts were opened and closed v. quickly, then that would be more of a concern. NB: that should even happen in a child query case - i.e. they should remain open.

I did think about using a QueueOf here instead of an array which would be a bit more efficient, but   wasn't that concerned.